### DISCUSSION TITLE
PnP.PowerShell Get-PnPList Error

### DISCUSSION THREAD

**Discussion by XxDaShTixX** [pnp/powershell#858] [Q&A]
We recently updated the **SharePointPnPPowerShellOnline** PowerShell module (as it is currency out of support according to Microsoft) to **PnP.PowerShell v1.6.0**. Ever since we updated, our runbooks intermittently / randomly fail on the cmdlet **Get-PnPList** with the following error: "_The data is not available. The query may not have been executed._" and I cannot pinpoint the reason to why this randomly fails.

$connection = Connect-PnPOnline -Credentials $creds -Url https://XXX.sharepoint.com/sites/YYY -ReturnConnection
$list = Get-PnPList -Identity 'Some List Name' -Connection $connection

Some additional information, this is running ion a schedule n Azure Automation on a hybrid worker. Note that the intermittent failures occur both when I run this in Azure Automation (hybrid worker) or run it locally (directly in ISE).

Anyone has this issue? Any suggestions? Thank you in advance for your help

**Comment by Sgalbus** [comment]
Same issue for me.
Does this problem has been solved?

**Reply by XxDaShTixX** [reply]
There was no solution provided to me for this, so far. One workaround was to build a new hybrid worker / server, but I wouldn't call this a solution as this isn't practical to have to build a new server when this error occurs after upgrading from SharePointPnPPowerShellOnline to PnP.PowerShell v1.6.0.

I hope this helps.

**Reply by AndersRask** [reply]
Could you please provide a bit more information.
Try
Get-pnpexeption
Or
Write out error in try/catch as
$_ | FL * -force

The old module have not been updated or changed so it is unlikely that this is what causes this. More likely a local issue on hybrid worker.

Did you manage to get it working on a runbook running on the automation account rather than the hybrid worker?

Also consider upgrading to pnp.powershell

**Reply by XxDaShTixX** [reply]
We already upgraded to PnP.PowerShell v1.6.0 as mentioned in the post. That is when the issue began to happen. We had no issues when working with the old SharePointPnPPowerShellOnline.

I had this in a try-catch block and the error it displayed is the error I quoted in the post.

I have had this issue with different hybrid workers and on my workstation. The only thing in common is that they all had the old SharePointPnPPowerShellOnline module installed - which I had removed before installing the new PnP.PowerShell module.

**Reply by AndersRask** [reply]
The error itself is not very descriptive, hence please provide a bit more information.

Try
Get-pnpexeption
Or
Write out error in try/catch as
$_ | FL * -force

This will give a full stack trace that might shed light on the issue.

Also be aware that running scripts in runbooks (both hybrid workers and "traditional" runbooks) is somewhat different from running PowerShell from your shell. I have had issues both with running jobs in parallel and jobs retrying due to invalid output from PnP (so always assign method calls like .Add() to a variable or use | Out-Null 

Running jobs in parallel where you close the connection in one job, can cause the other job to fail with the error you mention.

More info on runbook execution here:

https://docs.microsoft.com/en-us/azure/automation/automation-runbook-execution 

Some points that I usually share to my developers to avoid common traps when using runbooks:

Don’t use Write-Progress
if you use Write-Progress in an Azure Runbook, it will seriously slow your runbook down, increasing the cost, and if there are over 4000 write-progress calls the runbook will hang and crash.
Source: http://www.lieben.nu/liebensraum/2017/05/azure-runbooks-and-write-progress/ 

Write runbooks as idempotent with same input. Runbooks that are interrupted will restart from the beginning. This requires runbooks to be written in a way where they support restarts.

Restarts can be triggered by small things that can be difficult to identify, like a method on an object (eg $members.Add($user.Id) from PnP module ) that returns output to the flow. This can be circumvented by being very careful not to inadvertently return data to Output flow in the runbook. This means that method calls etc should always be caught in a variable or with | Out-Null unless output is ment to be returned.

Consider if runbook can handle to have multiple instances running of itself at the same time. Some runbooks may experience issues when running multiple jobs simultaniously. This can be avoided by checking for jobs of same type before executing the runbook:

```
$runningJobs = Get-AzAutomationJob <...> | ? {$_.Status -like "Running*" -or $_.Status -eq "New" -or $_.Status -eq "Queued" -Or $_.Status -eq "Starting"

if ($runningJobs.Count -gt 1 ) { Exit 1} else { #run job}
```
Be careful when using connections to other resources inside a runbook, since runbooks run in a shared sandbox! This means that you could risk affecting other runbooks, if you disconnect a connection. Hence connections should be created and disconnected explicitly. For example when doing PnP connections you should always specify the ReturnConnection switch parameter, and always use the connection specified, when calling a PnP cmdlet.
```
# begin block
$connection = Connect-PnPOnline -ReturnConnection <...> 
# process block
$groups = Get-PnPGroup -Connection $connection
# end block
if ($null -ne $connection) {
    Disconnect-PnPOnline -Connection $connection
}
```
Always use try/catch/[finally] for exception handling. Catch can be combined with throw or Write-Error "error" -ErrorAction Stop
NOTE: prefer the pattern catch { throw $_ } over catch { $PSCmdlet.ThrowTerminatingException($_)} unless you know exactly what you are doing, as TTE() only will result in a stream error, not an actual exception. This means the job will have status Completed and have stream errors inside, where if you use throw the job will have status failed! The reason is, that TTE() only is terminating the execution state (current function), not script state (runbook).

Hope this helps

**Reply by athltech** [reply]
I too have this same issue after upgrading from SharePointPnPPowerShellOnline to PnP.Powershell.  The issue seems to be that the Connect-PnPOnline command doesn't make a connection although it throws no exception.  If you run that command, it will complete without error.  Then, if you run Get-PnPSite, you will find that you are not connected to a site.  Any advice is appreciated.
